webauthn-rs-core/src/fake.rs

```diff
@@ -0,0 +1,323 @@
+                    // Android
+                    G_PIXEL_CRED_LEN
+                } else if type_dist < CRED_MULTI_WINDOWS {
+                    // Windows
+                    TPM_CRED_LEN
+                } else if type_dist < CRED_MULTI_APL {
+                    // Apple
+                    APPLE_CRED_LEN
+                } else if type_dist < CRED_MULTI_YK {
+                    // Other - yk
+                    YUBIKEY_CRED_LEN
+                } else {
+                    BITWARDEN_CRED_LEN
+                };
+
+                let mut cred = vec![0; cred_len];
+                seeded_rng.fill_bytes(&mut cred);
+
+                credentials.push(cred.into());
+            }
+        }
+
+        credentials
+    }
+}
+
+/// A fake `CredentialID` generator. This allows RP's to implement account enumeration defences if
+/// they so choose. Since webauthn requires the RP to send `CredentialID`s in challenges, when a
+/// user does not exist this will either provide an error or an empty list of IDs. This becomes a
+/// signal that an account with IDs associated to it in challenges must exist.
+///
+/// Account enumeration defence is a very subjective matter. For example on a website for a
+/// psychologists or drug addiction service, account enumeration prevention is important for
+/// privacy of the users of the service. For a service such as a business it may be less important
+/// as corporate directories may be available.
+///
+/// Because of this some considerations have been made in this api.
+///
+/// The only input we have from the account enumeration is the username.
+///
+/// Two users of this library should not have the same generated credentials given the same username.
+///
+/// `CredentialID`s that are generated should be believable and appear to come from legitimate services
+/// or credential sources. This includes realistic distribution of the types of devices that users
+/// may be using.
+///
+/// This api must be consistent - the same inputs will yield the same outputs.
+///
+/// To achieve this, the generator uses a seeded CSPRNG for each operation. The CSPRNG is seeded
+/// from the HMAC of the username. The HMAC is keyed from an input that the site provides.
+///
+/// The HMAC key should not be disclosed, as knowledge of the HMAC key will allow an external
+/// party to determine which IDs are generated and which are not.
+pub struct WebauthnFakeCredentialGenerator<D>
+where
+    D: FakeCredentialIDDistribution,
+{
+    // hmac key
+    hmac_key: pkey::PKey<pkey::Private>,
+    distribution: std::marker::PhantomData<D>,
+}
+
+impl<D> WebauthnFakeCredentialGenerator<D>
+where
+    D: FakeCredentialIDDistribution,
+{
+    /// Create a new HMAC keyed fake credential generator. You should associate a distribution type
+    /// using type annotations.
+    pub fn new(hmac_key: &[u8]) -> Result<Self, WebauthnError> {
+        let hmac_key = pkey::PKey::hmac(hmac_key).map_err(WebauthnError::OpenSSLError)?;
```

Comment:
There should be an explicit check that `hmac_key` is a "reasonable length". I feel like the 4 bytes used in the test isn't very long either 😅 

[Some versions of OpenSSL accept zero-length keys](https://github.com/openssl/openssl/issues/13089), and it seems like a very easy way to shoot yourself in the foot.

I'm also worried about a case where someone does something like:

```rust
let cred_gen: WebauthnFakeCredentialGenerator<FakePasskeyDistribution> =
            WebauthnFakeCredentialGenerator::new(b"www.example.com")?;
```

But I'm not sure if it's worth doing anything about it.

Reply:
We could always just have a small fn that returns a new random key.